Cjc/vp demo (#4320)
firedrakeproject/firedrake: demos/vlasov_poisson_1d/vp1d.py.rst

   \nabla_{\vec{x}} = (\partial_{x_1},\ldots, \partial_{x_d}), \quad
   \nabla_{\vec{v}} = (\partial_{v_1},\ldots, \partial_{v_d}).

To close the system, we need a formula for the acceleration :math:`\vec{a}`.
In the (single species) Vlasov-Poisson model, the acceleration is
determined by the electrostatic force,

.. math::
   \vec{a} = -\frac{1}{m}\nabla\phi,

where :math:`m`
is the mass per plasma particle, and :math:`\phi` is the electrostatic
potential determined by the Poisson equation,

.. math::
   -\nabla^2\phi = q_0\int_{\mathbb{R}^d} f(\vec{x},\vec{v},t)\,\mathrm{d} v,

where :math:`q_0` is the electric charge per plasma particle.

In this demo we specialise to :math:`d=1`, and the equations become

.. math::
   f_t + (fv)_x + (-f\phi_x/m)_v = 0, \quad
   -\phi_{xx} = q_0\int f(x,v,t)\,\mathrm{d} v,

with coordinates :math:`(x,v)\in \mathbb{R}^2`. From now on we will
relabel these coordinates :math:`(x,v)\mapsto (x_1,x_2)`, obtaining
the equivalent form,

.. math::
   f_t + \nabla\cdot(\vec{u}f) = 0, \quad \vec{u} = (v,-\phi_x/m), \quad
   -\phi_{x_1x_1} = q_0\int f(x_1,x_2,t)\,\mathrm{d} x_2,

where :math:`\nabla=(\partial_{x_1},\partial{x_2})`. From now we will

> choose units such that :math:`q_0,m` are absorbed into the definition of

Author: the latter.

each Runge-Kutta stage, we must solve for the electrostatic potential,
and then use it to compute :math:`\vec{u}`, in order to compute
:math:`\partial f/\partial t`.
   
As usual, to implement this problem, we start by importing the
Firedrake namespace. ::

  from firedrake import *

We build the mesh by constructing a 1D mesh, which will be extruded in
the vertical. Here we will use periodic boundary conditions in the
:math:`x_1` direction, ::
  
  ncells = 50
  L = 8*pi
  base_mesh = PeriodicIntervalMesh(ncells, L)

The mesh is then extruded upwards in the "velocity" direction. ::
  
  H = 10.0
  nlayers = 50
  mesh = ExtrudedMesh(base_mesh, layers=nlayers, layer_height=H/nlayers)

We want to have :math:`v=0` in the middle of the domain, so that we
can have negative and positive velocities. This requires to edit the
coordinate field. ::
		      
  Vc = mesh.coordinates.function_space()
  x, v = SpatialCoordinate(mesh)
  X = Function(Vc).interpolate(as_vector([x, v-H/2]))
  mesh.coordinates.assign(X)

Now we build a discontinuous finite element space for the density, ::
  

>   V = FunctionSpace(mesh, 'DG', 1)

Reviewer: We shouldn't be raising warnings in our own demos.
```suggestion
  V = FunctionSpace(mesh, 'DQ', 1)
```

specified through the ``vfamily``. ::
  
  Wbar = FunctionSpace(mesh, 'CG', 1, vfamily='R', vdegree=0)

We create a :class:`~.Function` to store the solution at the current
time, and then set its initial condition,

.. math::

   f(x,v,0) = \frac{1}{\sqrt{2\pi}}v^2\exp(-v^2/2)(1+ A\cos(kx)),
   \quad A=0.05, \quad k=0.5.

::
  
  fn = Function(V)
  A = Constant(0.05)
  k = Constant(0.5)
  fn.interpolate(v**2*exp(-v**2/2)*(1 + A*cos(k*x))/(2*pi)**0.5)

We will need the (conserved) average :math:`\bar{f}` for the Poisson
equation. ::

  One = Function(V).assign(1.0)
  fbar = assemble(fn*dx)/assemble(One*dx)

We create a :class:`~.Function` to store the electrostatic potential. ::

  phi = Function(Wbar)

The next task is to create the solver for the electrostatic potential, which
will be called every timestep. 
  
We create a :class:`~.Function` to store the right had side of the Poisson
equation. This will enable us to reuse the solver. ::

  fstar = Function(V)

Now we express the Poisson equation in UFL. ::
  
  psi = TestFunction(Wbar)
  dphi = TrialFunction(Wbar)
  phi_eqn = dphi.dx(0)*psi.dx(0)*dx - H*(fstar-fbar)*psi*dx

> To deal with :math:`\mathring{\bar{W}}`, we will precondition the
> problem with a shifted version, which is well-posed on :math:`\bar{W}`. ::
>   
>   shift_eqn = dphi.dx(0)*psi.dx(0)*dx + dphi*psi*dx
> 
> We use these to define a :class:`~.LinearVariationalProblem`. ::
>   
>   phi_problem = LinearVariationalProblem(lhs(phi_eqn), rhs(phi_eqn),
>                                          phi, aP=shift_eqn)
> 
> Now we build the :class:`~.LinearVariationalSolver`. The problem
> is preconditioned by the shifted operator which is solved using a direct
> solver, and we need to tell the solver about the nullspace of globally
> constant functions. ::
> 					 
>   params = {
>      'ksp_type': 'gmres',
>      'pc_type': 'lu',
>      'ksp_rtol': 1.0e-8,
>      }
>   nullspace = VectorSpaceBasis(constant=True)
>   phi_solver = LinearVariationalSolver(phi_problem,
>                                        nullspace=nullspace,
> 				       solver_parameters=params)

Reviewer: I've rearranged this a bit and added some more detail to make it more explicit how the null space is dealt with a) in the maths vs in the code, and b) how to remove the null space and how to solve the problem.

I'm not super attached to this specific rewording, so feel free to edit, but the original version was a bit vague about how Wbar, Wdotbar, and the shifted operator related to each other.

```suggestion
To deal mathematically with the null space of the potential, we expressed the
problem in :math:`\mathring{\bar{W}}`. Programmatically we will express the
problem in :math:`\bar{W}` and deal with the null space by defining a basis
for the space of globally constant functions, which we will later pass to PETSc
so the solver can remove it from the solution. ::

  nullspace = VectorSpaceBasis(constant=True)

However, the null space also means that the assembled matrix of the
Poisson problem will be singular, which will prevent us from using a
direct solver. To deal with this, we will precondition the Poisson problem
with a version shifted by :math:`\int_{\Omega}\phi\psi\mathrf{d}x`. The
shifted problem is well-posed on :math:`\bar{W}`, so the assembled matrix
will be non-singular and so solvable with direct methods. ::
  
  shift_eqn = dphi.dx(0)*psi.dx(0)*dx + dphi*psi*dx

We use these to define a :class:`~.LinearVariationalProblem`. ::
  
  phi_problem = LinearVariationalProblem(lhs(phi_eqn), rhs(phi_eqn),
                                         phi, aP=shift_eqn)

Now we build the :class:`~.LinearVariationalSolver`. The problem
is preconditioned by the shifted operator which is solved using a direct
solver, and we pass the nullspace of globally constant functions to
the solver. ::
					 
  params = {
     'ksp_type': 'gmres',
     'pc_type': 'lu',
     'ksp_rtol': 1.0e-8,
     }
  phi_solver = LinearVariationalSolver(phi_problem,
                                       nullspace=nullspace,
				       solver_parameters=params)
```

Now we move onto the solver to compute :math:`\partial f/\partial t`. We
define a symbolic :math:`\Delta t` which we will update later. ::
  
  dtc = Constant(0)

> The solver will take in ``fstar`` and return :math:`\Delta t\partial f/\partial t` in ``df_out``. ::

Reviewer: ```suggestion
At each stage, the solver will take in the intermediate solution ``fstar`` and
return the stage increment :math:`\Delta t\partial f/\partial t` in ``df_out``. ::
```

  dumpn = 0 # dump counter
  nsteps = 5000
  dt = T/nsteps
  dtc.assign(dt)

We set up some :class:`~.Function`\s to store Runge-Kutta stage variables. ::
  
  f1 = Function(V)
  f2 = Function(V)

We set up a ``VTKFile`` object to write output every ``ndump``
timesteps. ::

  outfile = VTKFile("vlasov.pvd")

We want to output the initial condition, so need to solve for the electrostatic
potential that corresponds to the initial density. ::

  fstar.assign(fn)
  phi_solver.solve()
  outfile.write(fn, phi)
  phi.assign(.0)

Now we start the timeloop using a lovely progress bar. ::

  for step in ProgressBar("Timestep").iter(range(nsteps)):

Each Runge-Kutta stage involves solving for :math:`\phi` before solving
for :math:`\partial f/\partial t`. Here is the first stage. ::

  #
      fstar.assign(fn)
      phi_solver.solve()
      df_solver.solve()
      f1.assign(fn + df_out)

The second stage. ::

  #
      fstar.assign(f1)
      phi_solver.solve()
      df_solver.solve()
      f2.assign(3*fn/4 + (f1 + df_out)/4)

The third stage. ::

  #
      fstar.assign(f2)
      phi_solver.solve()
      df_solver.solve()
      fn.assign(fn/3 + 2*(f2 + df_out)/3)

>       t += dt

Reviewer: Can this be written as a loop with the coefficient array defined beforehand? I don't think we want to be encouraging people to handcode unrolled Runge-Kutta loops.

Author: It's a bit fiddly to do that because of the triviality of the first stage, so it will all look a bit cryptic if I do that, and not very pedagogical. This is also how we did it in the original DG advection demo. I think it is harmless.

Author: Not that I can't use IRKsome because it isn't an IRK.

will be called every timestep. 
  
We create a :class:`~.Function` to store the intermediate densities at each
Runge-Kutta stage. The right hand side of the Poisson equation will be
evaluated using this :class:`~.Function` to obtain the potential at each
stage. Defining this beforehand will enable us to reuse the solver. ::

  fstar = Function(V)

Now we express the Poisson equation in UFL. ::
  
  psi = TestFunction(Wbar)
  dphi = TrialFunction(Wbar)
  phi_eqn = dphi.dx(0)*psi.dx(0)*dx - H*(fstar-fbar)*psi*dx

To deal mathematically with the null space of the potential, we expressed the
problem in :math:`\mathring{\bar{W}}`. Programmatically we will express the
problem in :math:`\bar{W}` and deal with the null space by defining a basis
for the space of globally constant functions, which we will later pass to PETSc
so the solver can remove it from the solution. ::

  nullspace = VectorSpaceBasis(constant=True, comm=COMM_WORLD)

However, the null space also means that the assembled matrix of the
Poisson problem will be singular, which will prevent us from using a
direct solver. To deal with this, we will precondition the Poisson problem

> with a version shifted by :math:`\int_{\Omega}\phi\psi\mathrf{d}x`. The

Reviewer 2: ```suggestion
with a version shifted by :math:`\int_{\Omega}\phi\psi\mathrm{d}x`. The
```
I think this is your issue. The log says
```
! Undefined control sequence.
l.17730 ...ifted by \(\int_{\Omega}\phi\psi\mathrf
                                                  {d}x\). The
```
